bugbug/models/bugtype.py

```diff
@@ -0,0 +1,102 @@
+# You can obtain one at http://mozilla.org/MPL/2.0/.
+
+import itertools
+
+import xgboost
+from imblearn.over_sampling import BorderlineSMOTE
+from sklearn.compose import ColumnTransformer
+from sklearn.feature_extraction import DictVectorizer
+from sklearn.pipeline import Pipeline
+from sklearn.multiclass import OneVsRestClassifier
+
+from bugbug import bug_features, bugzilla, feature_cleanup, labels
+from bugbug.model import BugModel
+import numpy as np
+
+
+class BugTypeModel(BugModel):
+    def __init__(self, lemmatization=False, historical=False):
+        BugModel.__init__(self, lemmatization)
+
+        self.sampler = BorderlineSMOTE(random_state=0)
+
+        feature_extractors = [
+            bug_features.has_str(),
+            bug_features.severity(),
+            # Ignore keywords that would make the ML completely skewed
+            # (we are going to use them as 100% rules in the evaluation phase).
+            bug_features.keywords({"regression", "talos-regression", "feature"}),
+            bug_features.is_coverity_issue(),
+            bug_features.has_crash_signature(),
+            bug_features.has_url(),
+            bug_features.has_w3c_url(),
+            bug_features.has_github_url(),
+            bug_features.whiteboard(),
+            bug_features.patches(),
+            bug_features.landings(),
+            bug_features.title(),
+            bug_features.blocked_bugs_number(),
+            bug_features.ever_affected(),
+            bug_features.affected_then_unaffected(),
+            bug_features.product(),
+            bug_features.component(),
+        ]
+
+        cleanup_functions = [
+            feature_cleanup.url(),
+            feature_cleanup.fileref(),
+            feature_cleanup.synonyms(),
+        ]
+
+        self.extraction_pipeline = Pipeline(
+            [
+                (
+                    "bug_extractor",
+                    bug_features.BugExtractor(feature_extractors, cleanup_functions),
+                ),
+                (
+                    "union",
+                    ColumnTransformer(
+                        [
+                            ("data", DictVectorizer(), "data"),
+                            ("title", self.text_vectorizer(min_df=0.001), "title"),
+                            (
+                                "first_comment",
+                                self.text_vectorizer(min_df=0.001),
+                                "first_comment",
+                            ),
+                            (
+                                "comments",
+                                self.text_vectorizer(min_df=0.001),
+                                "comments",
+                            ),
+                        ]
+                    ),
+                ),
+            ]
+        )
+
+        self.clf = OneVsRestClassifier(xgboost.XGBClassifier(n_jobs=16))
+
+    def get_labels(self):
+        classes = {}
+        keyword_list = []
```

Comment:
Why are you using all keywords as labels?

Reply:
Any value which comes in keywords would be a part of the label, right? (I need to group these according to their type, as multiple values can be of a single bug type.)

If not, is there a list of all available keywords that I should be using?

Reply:
There are keywords for a lot of disparate things, they don't only specify type. Also, some keywords can be grouped together to mean a single type.

There is an initial list of keywords in #20 (perf, memory, crash, keywords from https://wiki.mozilla.org/Security_Severity_Ratings).
We need to discover new ones we can use too, but that can be done later (iterate all bugs, put all keywords in a set, print them and see which ones we could use).

Reply:
For a start, should I implement only `crash` vs `perf` vs `memory` vs `security` in this PR? And we can keep adding others as we find them.

Reply:
Yes!